proof/refine/RISCV64/VSpace_R.thy

```diff
@@ -325,10 +332,39 @@ lemma deleteASIDPool_corres:
   apply clarsimp
   done
 
-crunch typ_at' [wp]: setVMRoot "\<lambda>s. P (typ_at' T p s)"
-  (simp: crunch_simps wp: crunch_wps)
+crunches unmapPageTable, unmapPage, setVMRoot, setMessageInfo, setMRs, performPageTableInvocation,
+         performASIDPoolInvocation, performPageInvocation
+  for typ_at' [wp]: "\<lambda>s. P (typ_at' T p s)"
+  and sc_at'_n[wp]: "\<lambda>s. P (sc_at'_n n p s)"
+  (wp: crunch_wps getASID_wp simp: crunch_simps)
+
+end
+
+sublocale Arch < unmapPageTable: typ_at_all_props' "unmapPageTable asid vaddr pt"
+  by typ_at_props'
 
-lemmas setVMRoot_typ_ats [wp] = typ_at_lifts [OF setVMRoot_typ_at']
+sublocale Arch < unmapPage: typ_at_all_props' "unmapPage magnitude asid vptr ptr"
+  by typ_at_props'
+
+sublocale Arch < setVMRoot: typ_at_all_props' "setVMRoot tcb"
+  by typ_at_props'
+
+sublocale Arch < setMessageInfo: typ_at_all_props' "setMessageInfo thread info"
+  by typ_at_props'
+
+sublocale Arch < setMRs: typ_at_all_props' "setMRs thread buffer messageData"
+  by typ_at_props'
+
+sublocale Arch < performPageTableInvocation: typ_at_all_props' "performPageTableInvocation iv"
+  by typ_at_props'
+
+sublocale Arch < performASIDPoolInvocation: typ_at_all_props' "performASIDPoolInvocation iv"
+  by typ_at_props'
+
+sublocale Arch < performPageInvocation: typ_at_all_props' "performPageInvocation iv"
```

Comment:
I'm curious, what does this do?

Reply:
The `typ_at_all_props'` locale will take the `typ_at'` and `sc_at'_n` lemmas we have and will generate all the `tcb_at'`, `ep_at'` type lemmas

Reply:
Cool, that's shiny new tech I'm not familiar with.